cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java

```diff
@@ -113,32 +277,138 @@ public Builder set(String fieldName, @Nullable Object value) {
+      if (localDate == null) {
+        fields.put(fieldName, null);
+        return this;
+      }
+      try {
+        fields.put(fieldName, Math.toIntExact(localDate.toEpochDay()));
+      } catch (ArithmeticException e) {
+        throw new IllegalArgumentException(String.format("Field %s is setting an invalid date", fieldName));
+      }
+      return this;
+    }
+
+    /**
+     * Sets the time value for {@link Schema.LogicalType#TIME_MILLIS} or {@link Schema.LogicalType#TIME_MICROS} field
+     *
+     * @param fieldName name of the field to set
+     * @param localTime value for the field
+     * @return this builder
+     * @throws UnexpectedFormatException if the field is not in the schema, or the field is not nullable but a null
+     *                                   value is given
+     */
+    public Builder setTime(String fieldName, @Nullable LocalTime localTime) {
+      Schema.Field field = validateAndGetField(fieldName, localTime);
+      Schema.LogicalType logicalType = validateAndGetLogicalType(fieldName, field, new HashSet<>(
+        Arrays.asList(Schema.LogicalType.TIME_MILLIS.name(), Schema.LogicalType.TIME_MICROS.name())));
+
+      if (localTime == null) {
+        fields.put(fieldName, null);
+        return this;
+      }
+
+      long nanos = localTime.toNanoOfDay();
+      if (logicalType == Schema.LogicalType.TIME_MILLIS) {
+        int millis = (int) TimeUnit.NANOSECONDS.toMillis(nanos);
+        fields.put(fieldName, millis);
+        return this;
+      }
+
+      long micros = TimeUnit.NANOSECONDS.toMicros(nanos);
+      fields.put(fieldName, micros);
+      return this;
+
+    }
+
+    /**
+     * Sets the timestamp value for {@link Schema.LogicalType#TIMESTAMP_MILLIS} or
+     * {@link Schema.LogicalType#TIMESTAMP_MICROS} field
+     *
+     * @param fieldName name of the field to set
+     * @param zonedDateTime value for the field
+     * @return this builder
+     * @throws UnexpectedFormatException if the field is not in the schema, or the field is not nullable but a null
+     *                                   value is given
+     * @throws IllegalArgumentException if the provided date is an invalid timestamp
+     */
+    public Builder setTimestamp(String fieldName, @Nullable ZonedDateTime zonedDateTime) {
+      Schema.Field field = validateAndGetField(fieldName, zonedDateTime);
+      Schema.LogicalType logicalType = validateAndGetLogicalType(fieldName, field, new HashSet<>(
+        Arrays.asList(Schema.LogicalType.TIMESTAMP_MILLIS.name(), Schema.LogicalType.TIMESTAMP_MICROS.name())));
+
+
+      if (zonedDateTime == null) {
+        fields.put(fieldName, null);
+        return this;
+      }
+
+      Instant instant = zonedDateTime.toInstant();
+      try {
+        if (logicalType == Schema.LogicalType.TIMESTAMP_MILLIS) {
+          long millis = Math.multiplyExact(instant.getEpochSecond(), 1000L);
+          long tsMillis = Math.addExact(millis, instant.getNano() / 1000000L);
```

Comment:
Use `TimeUnit.NANOSECONDS.toMillis` instead of manual division.

Reply:
`TimeUnit.NANOSECONDS.toMillis();` divides by `1000000L` instead of `1000L`. To divide by `1000L` we can use `TimeUnit.NANOSECONDS.toMicros()` which is kind of confusing. Thats why manual division.

Reply:
? I don't get it. Isn't that you are dividing by `100000L` here?